#!/bin/bash

#######################
### This script is an illustration of the genetic screening for Cystic Fibrosis(CF).

### Usage: ./CFscreen.sh 

########
### Goal: use genomic data from a patient to identify CF-relevant mutations

########
### Main steps:

# 1. Identify mutations in CFTR gene present in HG00121 individual from 1000 Genomes project 
# Sample info: http://www.internationalgenome.org/data-portal/sample/HG00121 
# Sequence Data: ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/HG00121/

# 2. Check if any identified mutations are CF-causing based on the list provided by CFTR2 website (https://www.cftr2.org/mutations_history)    

# 3. Compare results with the results recorded in VCF file generated by 1000 Genomes project
# VCF data: http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/



####################### Software used #################
# samtools 1.8-1-gc814890
# bcftools 1.8-8-g4119a24
# ensembl-vep 92.0


####################### Data files in Data folder ##########

### human reference genome GRCh37
# wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/phase2_reference_assembly_sequence/hs37d5.fa.gz

### List of CFTR mutations provided by https://www.cftr2.org/
# https://www.cftr2.org/sites/default/files/CFTR2_8December2017_2.xlsx
# saved as CFTR2_list.csv

### Chromosome 7 variants from 1000 Genomes project (VCF)
# wget -O All.chr7.vcf.gz http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr7.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz 

### .tbi Index file for Chromosome 7 vcf
# wget -O All.chr7.vcf.gz.tbi ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr7.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz.tbi


########################################################
####################### Analysis #######################

inRef="Data/hs37d5.fa.gz"        #reference genome referenced in bam file
inCFTR2="Data/CFTR2_list.csv"    #CFTR mutations from https://www.cftr2.org/
inVCF1000="Data/All.chr7.vcf.gz" #VCF file for chr7 from 1000 genomes project
out="Results.txt"
VEP="$HOME/ensembl-vep"          #path to VEP=Variant Effect Predictor



########### Get patient's data (sample HG00121) ########

### The starting point of the analysis is the alignment file (bam) generated by aligning HG00121 whole genome sequence data to human reference genome using BWA aligner.The exact steps of generating the alignment can be found here: ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/historical_data/former_toplevel/README.alignment_data.md As these steps are standard proceedure for generating bam files for various projects in genomics, I am saving time here and pretend that I made the alignment using raw data available here: ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/HG00121/sequence_read/ 

#Get low coverage alignment for CFTR region: 117105838-117356025(GRCh37 )
#https://useast.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000001626;r=7:117465784-117715971

samtools view -bh ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/HG00121/alignment/HG00121.mapped.ILLUMINA.bwa.GBR.low_coverage.20130415.bam 7:117105838-117356025 > CFTR.bam

# view CFTR.bam header to identify the source of reference sequence (needed to call variants)
#samtools view -H CFTR.bam

# relevant samtools view -H CFTR.bam output:
# @SQ	SN:7	LN:159138663	M5:618366e953d6aaad97dbe4777c29375e	UR:ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/phase2_reference_assembly_sequence/hs37d5.fa.gz	AS:NCBI37	SP:Human



############# Variant calling and filtering #############

# Call variants with bcftools 
bcftools mpileup -f $inRef CFTR.bam | bcftools call -mv -Oz -o CFTR_HG00121.vcf.gz

# Check number of called variants and quality of the calls (total: 76)
echo "Number of variants called from CFTR.bam : $(zcat CFTR_HG00121.vcf.gz | grep -v '^#' | cut -f6 | sort -n | wc -l)" >> $out

# Index and filter based on quality (QUAL >=40)
bcftools index -t CFTR_HG00121.vcf.gz

bcftools filter -i 'QUAL>=40' CFTR_HG00121.vcf.gz -Oz -o CFTR_HG00121_QUAL40.vcf.gz

# Check number of called variants after quality filter ( total: 50)
echo "Number of variants after quality filter (QUAL >=40): $(zcat CFTR_HG00121_QUAL40.vcf.gz | grep -v '^#' | cut -f6 | sort -n | wc -l)" >> $out


############# Variant annotation and filtering ##########

# using Variant Effect Predictor Tool from Ensembl 
# https://useast.ensembl.org/info/docs/tools/vep/index.html

mkdir -p VEP

# get all information available
$VEP/vep -i CFTR_HG00121_QUAL40.vcf.gz -o VEP/CFTR_HG00121.txt --cache --dir_cache $HOME/.vep/homo_sapiens_37 --port 3337 --everything --force_overwrite

echo "VEP results for CFTR_HG00121_QUAL40.vcf: VEP/CFTR_HG00121.txt; VEP/CFTR_HG00121.txt_summary.html " >> $out

# use filter_vep to filter results as you wish
# https://useast.ensembl.org/info/docs/tools/vep/script/vep_filter.html

# Here we are intersted if any of the detected variants are implicated in cystic fibrosis
# Known variants are available at https://www.cftr2.org/mutations_history 
# Select rsIDs we need to check for:

awk -F"," '$4~/rs[^I]/ {print $4,$9}' OFS="\t" $inCFTR2 | \
    grep -v 'Non CF-causing' | \
    cut -f1| awk -F" " '{print $1}'> CFTR2_CF-relevant_list.txt 

# filter initial VEP output to select rsIDs from CFTR2_CF-relevant_list.txt
$VEP/filter_vep -i VEP/CFTR_HG00121.txt -o VEP/CFTR_HG00121_CF-relevant.txt -filter "Existing_variation in CFTR2_CF-relevant_list.txt" --force_overwrite

echo "VEP results for CFTR_HG00121_QUAL40.vcf with CF-relevant variants: VEP/CFTR_HG00121_CF-relevant.txt " >> $out


################# RESULTS ################

echo -e "\nRESULTS for CFTR_HG00121_QUAL40.vcf:
HG00121 individual does not have variants implicated in cystic fibrosis." >> $out

######################################################################
############ COMPARE WITH AVAILABLE VCF from 1000 Genomes ############

#Get CFTR region from available VCF
#bcftools index -t $inVCF1000

bcftools view -r 7:117105838-117356025 -s HG00121 $inVCF1000 -Oz -o CFTR_HG00121_1000genomes.vcf.gz

#get het or homo for ALT
zcat CFTR_HG00121_1000genomes.vcf.gz |grep -E -v  '0\|0' |bgzip -c > CFTR_HG00121_1000genomes_VAR.vcf.gz

# Check number of called variants in CFTR_HG00121_1000genomes_VAR.vcf.gz (106)
echo "Number of variants called in CFTR_HG00121_1000genomes_VAR.vcf: $(zcat CFTR_HG00121_1000genomes_VAR.vcf.gz | grep -v '^#' |  wc -l)" >> $out

###
#check if CFTR_HG00121_QUAL40.vcf.gz is a subset of CFTR_HG00121_1000genomes.vcf.gz

zcat CFTR_HG00121_QUAL40.vcf.gz|grep -v '^#'|cut -f2 > CFTR_HG00121_QUAL40_pos.txt
zcat CFTR_HG00121_1000genomes_VAR.vcf.gz |grep -v '^#'|cut -f2 > CFTR_HG00121_1000genomes_VAR_pos.txt

#positions common between 2 files (49)
CommonVar=$(comm -1 -2 CFTR_HG00121_QUAL40_pos.txt CFTR_HG00121_1000genomes_VAR_pos.txt|wc -l)
echo "There are $CommonVar common variants between CFTR_HG00121_QUAL40.vcf and CFTR_HG00121_1000genomes_VAR.vcf" >> $out

#position unique to CFTR_HG00121_QUAL40_pos.txt
echo "$(comm -2 -3 CFTR_HG00121_QUAL40_pos.txt CFTR_HG00121_1000genomes_VAR_pos.txt) \
is unique variant position to CFTR_HG00121_QUAL40.vcf.gz, not present in CFTR_HG00121_1000genomes_VAR.vcf" >> $out

#check VEP annotations for CFTR_HG00121_1000genomes_VAR.vcf.gz
$VEP/vep -i CFTR_HG00121_1000genomes_VAR.vcf.gz -o VEP/CFTR_HG00121_1000genomes.txt --cache --dir_cache $HOME/.vep/homo_sapiens_37 --port 3337 --everything --force_overwrite

echo "VEP results for CFTR_HG00121_1000genomes_VAR.vcf: VEP/CFTR_HG00121_1000genomes.txt; VEP/CFTR_HG00121_1000genomes.txt_summary.html " >> $out

#check if any of CF-relevant mutations are present in 1000 Genomes data HG00121
$VEP/filter_vep -i VEP/CFTR_HG00121_1000genomes.txt -o VEP/CFTR_HG00121_1000genomes_CF-relevant.txt -filter "Existing_variation in CFTR2_CF-relevant_list.txt" --force_overwrite

echo "VEP results for CFTR_HG00121_1000genomes_VAR.vcf with CF-relevant variants: VEP/CFTR_HG00121_1000genomes_CF-relevant.txt" >> $out

################# RESULTS ################

echo -e "\nRESULTS for CFTR_HG00121_1000genomes_VAR.vcf
HG00121 individual does not have variants implicated in cystic fibrosis." >> $out

################ CONCLUSION #####################

echo -e  "\nCONCLUSIONS: 
There are 50 variants (QUAL >=40) in CFTR_HG00121_QUAL40.vcf and 106 variants in CFTR_HG00121_1000genomes_VAR.vcf
One of the variants present in CFTR_HG00121_QUAL40.vcf is not found in CFTR_HG00121_1000genomes_VAR.vcf
These differences are likely the result of 
1) differences in the data analyzed (CFTR_HG00121_QUAL40.vcf is based on the subset of data used to generate CFTR_HG00121_1000genomes_VAR.vcf)
2) differences in variant calling tools
(The Genomes Project C. (2015). A global reference for human genetic variation. Nature, 526, 68. doi: 10.1038/nature15393)" >> $out

cat $out